A43-sdk-authorization.md

```diff
@@ -0,0 +1,364 @@
+	  "scoped by services.",
+	"type": "object",
+	"properties": {
+		"name": {
+			"description": "The name of an authorization policy. It is mainly for monitoring and error message generation.",
+			"type": "string",
+		},
+		"deny_rules": {
+			"description": "List of deny rules to match. If a request matches any of the deny rules, then it will be denied. If none of"
+			  "the deny rules matches or there are no deny rules, the allow rules will be evaluated.",
+			"type": "array",
+			"items": {
+				"$ref": "#/definitions/rule",
+			},
+		},
+		"allow_rules": {
+			"description": "List of allow rules to match. The allow rules will only be evaluated after the deny rules. If a request"
+			  "matches any of the allow rules, then it will allowed. If none of the allow rules matches, it will be denied.",
+			"type": "array",
+			"items": {
+				"$ref": "#/definitions/rule",
+			},
+		}
+	},
+	"required": ["name", "allow_rules"]
+}
+```
+
+gRPC SDK authorization policy has a list of allow rules and a list of deny rules. 
+The following sequence is followed to make an authorization decision -
+1. Check for a match in deny rules, if a match is found, the request will be denied. 
+   If no match is found in deny rules, or there are no deny rules execute next step.
+2. Check for a match in allow rules, if a match is found, the request will be allowed. 
+   Note that allow rules is a required field here, if not present, policy is invalid.
+3. If no match is found in allow rules, we deny the request.
+
+In the following policy example
+- Peer identity from ["spiffe://foo.com/sa/admin1", "spiffe://foo.com/sa/admin2"] is 
+  authorized to access any RPC methods in pkg.service
+- Any authenticated user is allowed to access "foo" and "bar" RPC methods.
+- Nobody can access "secret" RPC method, not even the admins as deny rules are evaluated 
+  first.
+
+```json
+{
+	"name": "example-policy",
+	"allow_rules": [
+		{
+			"name": "admin-access",
+			"source": {
+				"principals": [
+					"spiffe://foo.com/sa/admin1",
+					"spiffe://foo.com/sa/admin2"
+				]
+			},
+			"request": {
+				"paths": ["/pkg.service/*"]
+			}
+		},
+		{
+			"name": "dev-access",
+			"source": {
+				"principals": [
+					"*"
+				]
+			},
+			"request": {
+				"paths": [
+					"/pkg.service/foo",
+					"/pkg.service/bar"
+				]
+			}
+		}
+	],
+	"deny_rules": [
+		{
+			"name": "deny-access",
+			"request": {
+				"paths": [
+					"*/secret"
+				]
+			}
+		}
+	]
+}
+```
+
+As mentioned previously, internally gRPC Authorization engine consumes Envoy RBAC 
+proto. The user provided JSON SDK authorization policy will be translated to Envoy 
+RBAC proto and provided to Engines for evaluation. Note that SDK authorization policy 
+is a subset of Envoy RBAC, and it does not support all the fields that are present in 
+Envoy RBAC.
+
+### API
+
+gRPC will support both static initialization and dynamically reloading the policy
```

Comment:
It should be mentioned that applying more than one policy (of any kind) is not allowed (and why - because the semantics are not define and/or implementation). And how this will be detected and prevented.

Reply:
That's right! So we should accept only a single SDK authorization policy per gRPC Server.
Since we do deny by default in SDK path, if we have multiple SDK policies, say different policies for different services in a server, we would only evaluate against the first policy in chain if no match were found.

Only C++, currently ensures this behavior, where we accept only a single policy.
In Java and Go, since we expose the interceptors, users could create multiple sdk interceptors (therefore multiple policies) per server.  Should we be updating the design to check and prevent this, or just state explicitly that only one policy per server?

Reply:
> ...Should we be updating the design to check and prevent this, or just state explicitly that only one policy per server?

Is it easy/possible to detect and prevent? What does it take?

Reply:
Based on our email discussion, we won't be updating the current designs.  It is the user's responsibility to inject compatible policies when using multiple interceptors. I included a paragraph in proposal recommending single policy per grpc server.